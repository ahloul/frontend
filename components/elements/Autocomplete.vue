<template>
  <div class="relative">
    <!-- Name INPUT -->
    <input
      v-model="selection[context.attributes.dname]"
      v-bind="context.attributes"
      class="form-input"
      type="search"
      autocomplete="off"
      @blur="handleBlur"
      @input="change"
    />

    <ul v-if="openSuggestion" class="dropdown w-full">
      <li
        v-for="(suggestion, index) in list"
        :key="index"
        class="dropdown-item"
        @click="suggestionClick(suggestion)"
      >
        {{ suggestion[context.attributes.dname] }}
      </li>
    </ul>
  </div>
</template>

<script>
import { debounce, clone, isEmpty } from 'lodash'

export default {
  name: 'Autocomplete',
  props: {
    context: {
      type: Object,
      // required: true,
      default: () => {
        return {
          attributes: {
            endpoint: '',
            dname: 'name',
            placeholder: '',
            label: '',
          },
        }
      },
    },
  },
  data: () => ({
    open: false,
    selection: {
      name: null,
    },
    list: [],
  }),
  computed: {
    // The flag
    openSuggestion() {
      return this.open === true
    },
  },
  beforeMount() {
    this.selection[this.context.attributes.dname] =
      this.context.model.display || {}
  },
  mounted() {
    // if (isEmpty(this.initial)) return
    this.selection = this.context.model || {}
  },
  methods: {
    // Refresh the list when the user changes input
    change: debounce(async function (e) {
      const { endpoint, queryname, dname } = this.context.attributes
      if (!this.open) this.open = true
      const data = await this.$axios.$get(
        `/api/${endpoint}`,
        this.selection[dname]
          ? {
              params: {
                [queryname]: this.selection[dname],
                limit: 5,
              },
            }
          : null
      )
      this.list = data?.rows || data
    }, 750),

    // When one of the suggestion is clicked
    suggestionClick(selection) {
      if (isEmpty(selection)) return
      const { address, label, locationId } = selection
      this.context.model = { ...address, label, locationId }
      this.selection = clone(selection)
      this.open = false
    },

    handleBlur: debounce(function () {
      this.open = false

      this.context.blurHandler()
      if (!this.selection[this.context.attributes.dname])
        return (this.context.model = '')
    }, 200),
  },
}
</script>

<style lang="scss" scoped>
.dropdown {
  @apply bg-white shadow rounded-lg mt-1 absolute z-30;
  &-item {
    @apply p-3 text-sm w-full;
    &:hover {
      @apply bg-grey cursor-pointer;
    }
  }
}
</style>
