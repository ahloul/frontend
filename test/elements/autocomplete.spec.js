import { mount, shallowMount } from '@vue/test-utils'
import Autocomplete from '../../components/elements/Autocomplete.vue'

describe('Autocomplete', () => {
  let wrapper

  test('is a Vue instance', () => {
    wrapper = mount(Autocomplete)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })
  test('computed openSuggestion function', () => {
    const cmp = mount(Autocomplete, {
      data: () => ({
        open: false,
      }),
    })
    expect(cmp.vm.$data.open).toBe(false)
    expect(cmp.find('li').exists()).toEqual(false)
    cmp.setData({ open: true })
    expect(cmp.vm.$data.open).toBe(true)
  })
  test('emitted selection', async () => {
    wrapper = mount(Autocomplete)
    wrapper.vm.$emit('selection', 'start')
    await wrapper.vm.$nextTick()
    expect(wrapper.emitted().selection[0]).toEqual(['start'])
  })
  //  test('li emitted suggestion', () => {
  //    wrapper = shallowMount(Autocomplete)
  //    wrapper.find('input').trigger('input')
  //    expect(wrapper.vm.input).toHaveBeenCalled()
  //  })
  // test('input change', () => {
  //    const initialValue = 'Initial value'
  //    wrapper = shallowMount(Autocomplete, { propsData: { initialValue } })

  //    const newValue = 'updated value'
  //    const input = wrapper.find('input')
  //    (wrapper.find('input').element as HTMLInputElement).value =newValue
  //    input.trigger('change')
  //   expect(wrapper.text()).toContain(newValue)
  //  })
})
