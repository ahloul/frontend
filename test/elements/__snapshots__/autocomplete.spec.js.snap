// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete snapshot follow the user through the form 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;

exports[`Autocomplete snapshot follow the user through the form 2`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;
